_docs/tasks/integrating-services-into-istio.md

```diff
@@ -1,36 +1,224 @@
+```
+
+
+The Envoy sidecar does _not_ intercept container-to-container traffic
+within the same pod when traffic is routed via localhost. This is by
+design.
 
+```bash
+$ kubectl exec -it ${SERVER} -c echo -- curl localhost:8080 | grep x-request-id
+```
+
+## Understanding what happened
+
+`istioctl kube-inject` injects additional containers into YAML
+resource on the client _before_ submitting to the Kubernetes API
+server. This will eventually be replaced by server-side injection via
+admission controller. Use `kubectl get deployment echo -o yaml` to
+inspect the modified deployment and look for the following:
+
+* A proxy container which includes the Envoy proxy and agent to manage
+  local proxy configuration.
+
+* An [init-container](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)
+  to program [iptables](https://en.wikipedia.org/wiki/Iptables).
+
+The proxy container includes the Envoy proxy and corresponding agent
+to manage the proxy configuration. The sidecar runs with a specific
+UID so that the iptables can differentiate outbound traffic from the
+proxy itself from the applications which are redirected to proxy.
+
+```yaml
+- args:
+    - proxy
+    - sidecar
+    - "-v"
+    - "2"
+  env:
+    -
+      name: POD_NAME
+      valueFrom:
+        fieldRef:
+          apiVersion: v1
+          fieldPath: metadata.name
+    -
+      name: POD_NAMESPACE
+      valueFrom:
+        fieldRef:
+          apiVersion: v1
+          fieldPath: metadata.namespace
+    -
+      name: POD_IP
+      valueFrom:
+        fieldRef:
+          apiVersion: v1
+          fieldPath: status.podIP
+  image: "docker.io/istio/proxy:<...tag... >"
+  imagePullPolicy: Always
+  name: proxy
+  securityContext:
+    runAsUser: 1337
+
+```
+
+iptables is used to transparently redirect all inbound and outbound
+traffic to the proxy. An init-container is used for two reasons:
+
+1. iptables requires
+[NET_CAP_ADMIN](http://man7.org/linux/man-pages/man7/capabilities.7.html).
+
+2. The sidecar iptable rules are fixed and don't need to be updated
+after pod creation. The proxy container is responsible for dynamically
+routing traffic.
+
+```json
+pod.beta.kubernetes.io/init-containers: '[
+  {
+    "name":"init",
+    "image":"docker.io/istio/init:<..tag...>",
+    "args":[ "-p", "15001", "-u", "1337" ],
+    "imagePullPolicy":"Always",
+    "securityContext":{
+      "capabilities":{
+        "add":[
+          "NET_ADMIN"
+        ]
+      }
+    }
+  },
+  {
+    "name":"enable-core-dump",
+    "image":"alpine",
+    "command":[ "/bin/sh"],
+    "args":[ "-c", "sysctl -w kernel.core_pattern=/tmp/core.%e.%p.%t \u0026\u0026 ulimit -c unlimited" ],
+    "imagePullPolicy":"Always",
+    "securityContext":{
+      "privileged":true
+    }
+  }
+]'
+```
+
+The `enable-core-dump` init-container is also added to enable GDB core
+dumps for the Envoy proxy. This can be disabled by passing
+`--coreDump=False` to `istioctl kube-inject`.
```

Comment:
We should remove the core dump part. This is not very comforting to an end user 😉